external-crates/move/crates/move-regex-borrow-graph/src/references.rs

```diff
@@ -0,0 +1,436 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    bail,
+    collections::{Path, Paths},
+    error,
+    regex::Regex,
+    Result,
+};
+use itertools::Either;
+use std::{
+    cmp,
+    collections::{BTreeMap, BTreeSet},
+    fmt::{self, Debug},
+};
+
+//**************************************************************************************************
+// Definitions
+//**************************************************************************************************
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Ref(Ref_);
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
+enum Ref_ {
+    Canonical(usize),
+    Fresh(usize),
+}
+
+#[derive(Clone)]
+struct LocRegex<Loc, Lbl> {
+    loc: Loc,
+    regex: Regex<Lbl>,
+}
+
+#[derive(Clone, Debug)]
+struct Edge<Loc, Lbl: Ord> {
+    abstract_size: usize,
+    regexes: BTreeSet<LocRegex<Loc, Lbl>>,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Node<Loc, Lbl: Ord> {
```

Comment:
Is it actually preferable to manage our own graph as opposed to using a graph library with edge and node annotations?

Reply:
Given some of the strange things I need to do around self loops, I think so?